puan/logic/plog/__init__.py

```diff
@@ -355,41 +360,138 @@ def errors(self) -> typing.List[PropositionValidationError]:
             -------
                 out : List[:class:`PropositionValidationError`]
         """
-        errors = []
-        deps = self._dependencies()
-        try:
-            list(
-                toposort.toposort(
-                    dict(deps),
-                )
-            )
-        except toposort.CircularDependencyError:
-            errors.append(PropositionValidationError.CIRCULAR_DEPENDENCIES)
-        except Exception as e:
-            raise Exception(f"got unhandled error: {e}")
+        return list(
+            maz.compose(
+                functools.partial(
+                    itertools.compress,
+                    [
+                        PropositionValidationError.CIRCULAR_DEPENDENCIES,
+                        PropositionValidationError.AMBIVALENT_VARIABLE_DEFINITIONS,
+                        PropositionValidationError.AMBIVALENT_VARIABLE_DEFINITIONS,
+                        PropositionValidationError.NON_UNIQUE_SUB_PROPOSITION_SET
+                    ],
+                ),
+                maz.fnmap(
 
-        # covers (a -> a) case, when there's a direct circular dependency.
-        # this is not covered by `toposort` lib
-        if any(itertools.starmap(lambda a,b: a in b, deps)):
-            errors.append(PropositionValidationError.CIRCULAR_DEPENDENCIES)
+                    # Checks that if there exists any circlular
+                    # variable dependecy somewhere in the model
+                    maz.fnexcept(
+                        maz.compose(
+                            operator.not_,
+                            functools.partial(
+                                operator.eq,
+                                None,
+                            ),
+                            operator.methodcaller("prepare"),
+                            graphlib.TopologicalSorter,
```

Comment:
How is this different from the old toposort? (Just curious)

Reply:
This one is actually python's package and it also fails for these cases A -> A (which it should and the old one didn't)